File: mesh/Polyhedron/Polyhedron/volint.cpp
//
//  volume.cpp
//  
//
//  Created by Yang Yu on 14-11-14.
//
//

#include "volint.h"

void VolumeIntegrals(POLYHEDRON &p)
{
    int i, j, a, b, c;
    double nx, ny, nz, tmp;
    double *Wf, **Nf;
    double F[12], wf, nf[3];
    double VolIntZeroth, VolIntFirst[3], VolIntSecond[6];
    
    VolIntZeroth = 0.0;
    for(i=0; i<3; i++) VolIntFirst[i] = 0.0;
    for(i=0; i<6; i++) VolIntSecond[i] = 0.0;
    
    CompWfNf(p, Wf, Nf);
    
    for (i=0; i<p.NumFaces; i++)
    {
        wf = Wf[i];
        for(j=0; j<3; j++) nf[j] = Nf[i][j];

        nx = fabs(nf[0]);
        ny = fabs(nf[1]);
        nz = fabs(nf[2]);
        
        if ((nx>ny) && (nx>nz)) c = 0;
        else if (ny>nz) c = 1;
        else c = 2;
        
        a = (c + 1) % 3;
        b = (a + 1) % 3;
        
        FaceIntegrals(p, wf, nf, i, a, b, c, F);
        
        if (a==0) tmp = F[0];
        else if (b==0) tmp = F[1];
        else tmp = F[2];
            
        VolIntZeroth = VolIntZeroth + Nf[i][0] * tmp;
        
        VolIntFirst[a] = VolIntFirst[a] + Nf[i][a] * F[3];
        VolIntFirst[b] = VolIntFirst[b] + Nf[i][b] * F[4];
        VolIntFirst[c] = VolIntFirst[c] + Nf[i][c] * F[5];
        
        VolIntSecond[a] = VolIntSecond[a] + Nf[i][a] * F[6];
        VolIntSecond[b] = VolIntSecond[b] + Nf[i][b] * F[7];
        VolIntSecond[c] = VolIntSecond[c] + Nf[i][c] * F[8];
        VolIntSecond[a+3] = VolIntSecond[a+3] + Nf[i][a] * F[9];
        VolIntSecond[b+3] = VolIntSecond[b+3] + Nf[i][b] * F[10];
        VolIntSecond[c+3] = VolIntSecond[c+3] + Nf[i][c] * F[11];
    }
    
    for(j=0; j<3; j++)
    {
        VolIntFirst[j] = VolIntFirst[j] / 2.0;
        VolIntSecond[j] = VolIntSecond[j] / 3.0;
        VolIntSecond[j+3] = VolIntSecond[j+3] / 2.0;
    }
    
    VolumeInfoCalcu(VolIntZeroth, VolIntFirst, VolIntSecond, p);
    
}

//
//  Compute the Wf and Nf
//

void CompWfNf(POLYHEDRON &p, double *&Wf, double **&Nf)
{
    int i, j;
    Vector ra, rb, rc, rab, rbc, tmpv;
    
    Wf = (double *) malloc(p.NumFaces*sizeof(double));
    Nf = (double **) malloc(p.NumFaces*sizeof(double *));
    for (i=0; i<p.NumFaces; i++) Nf[i] = (double *) malloc(3*sizeof(double));
    
    for(i=0; i<p.NumFaces; i++)
    {
        vectorSet(ra, p.Vertices[p.Faces[i][0]][0], p.Vertices[p.Faces[i][0]][1], p.Vertices[p.Faces[i][0]][2]);
        vectorSet(rb, p.Vertices[p.Faces[i][1]][0], p.Vertices[p.Faces[i][1]][1], p.Vertices[p.Faces[i][1]][2]);
        vectorSet(rc, p.Vertices[p.Faces[i][2]][0], p.Vertices[p.Faces[i][2]][1], p.Vertices[p.Faces[i][2]][2]);
        
        vectorSub(ra,rb,rab);
        vectorSub(rb,rc,rbc);
        
        vectorCross(rab, rbc, tmpv);
        vectorNorm(tmpv);
        
        Wf[i] = - vectorDot(ra, tmpv);
        
        for(j=0; j<3; j++) Nf[i][j] = tmpv[j];
    }
    
}

//
//  Compute the face integrals
//

void FaceIntegrals(POLYHEDRON &p, double wf, double nf[3], int id, int a, int b, int c, double F[12])
{
    double P[10], K[4];
    
    ProjectionIntegral(p, a, b, id, P);
    
    K[0] = 1.0 / nf[c];
    K[1] = K[0] * K[0];
    K[2] = K[1] * K[0];
    K[3] = K[2] * K[0];
    
    F[0] = K[0] * P[1];
    F[1] = K[0] * P[2];
    F[2] = - K[1] * (nf[a] * P[1] + nf[b] * P[2] + wf * P[0]);
    
    F[3] = K[0] * P[3];
    F[4] = K[0] * P[5];
    F[5] = K[2] * (nf[a] * nf[a] * P[3] + 2.0 * nf[a] * nf[b] * P[4] + \
                   nf[b] * nf[b] * P[5] + wf * (2.0 * (nf[a] * P[1] + \
                   nf[b] * P[2]) + wf * P[0]));
    
    F[6] = K[0] * P[6];
    F[7] = K[0] * P[9];
    F[8] = - K[3] * (nf[a] * nf[a] * nf[a] * P[6] + 3.0 * nf[a] * nf[a] * nf[b] * P[7] + \
                     3.0 * nf[a] * nf[b] * nf[b] * P[8] + nf[b] * nf[b] * nf[b] * P[9] + \
                     3.0 * wf * (nf[a] * nf[a] * P[3] + 2.0 * nf[a] * nf[b] * P[4] + \
                     nf[b] * nf[b] * P[5]) + wf * wf * (3.0 * (nf[a] * P[1] + nf[b] * P[2]) + \
                     wf * P[0]));
                     
    F[9] = K[0] * P[7];
    F[10] = - K[1] * (nf[a] * P[8] + nf[b] * P[9] + wf * P[5]);
    F[11] = K[2] * (nf[a] * nf[a] * P[6] + 2.0 * nf[a] * nf[b] * P[7] + nf[b] * nf[b] * P[8] + \
                    wf * (2.0 * (nf[a] * P[3] + nf[b] * P[4]) + wf * P[1]));
                     
}

//
//  Compute the projection integral
//

void ProjectionIntegral(POLYHEDRON &p, int a, int b, int id, double P[10])
{
    int i;
    double A0, A1, DA;
    double B0, B1, DB;
    double A02, A03, A04, B02, B03, B04;
    double A12, A13, B12, B13;
    double C1, CA, CAA, CAAA, CB, CBB, CBBB;
    double CAB, KAB, CAAB, KAAB, CABB, KABB;
    
    for(i=0; i<10; i++) P[i] = 0.0;
    
    for (i=0; i<3; i++)
    {
        A0 = p.Vertices[p.Faces[id][i]][a];
        B0 = p.Vertices[p.Faces[id][i]][b];
        A1 = p.Vertices[p.Faces[id][(i+1) % 3]][a];
        B1 = p.Vertices[p.Faces[id][(i+1) % 3]][b];
        
        DA = A1 - A0;
        DB = B1 - B0;
        
        A02 = A0 * A0;
        A03 = A02 * A0;
        A04 = A03 * A0;
        
        B02 = B0 * B0;
        B03 = B02 * B0;
        B04 = B03 * B0;
        
        A12 = A1 * A1;
        A13 = A12 * A1;
        B12 = B1 * B1;
        B13 = B12 * B1;
        
        C1 = A1 + A0;
        CA = A1 * C1 + A02;
        CAA = A1 * CA + A03;
        CAAA = A1 * CAA + A04;
        CB = B1 * (B1 + B0) + B02;
        CBB = B1 * CB + B03;
        CBBB = B1 * CBB + B04;
        CAB = 3.0 * A12 + 2.0 * A1 * A0 + A02;
        KAB = A12 + 2.0 * A1 * A0 + 3.0 * A02;
        CAAB = A0 * CAB + 4.0 * A13;
        KAAB = A1 * KAB + 4.0 * A03;
        CABB = 4.0 * B13 + 3.0 * B12 * B0 + 2.0 * B1 * B02 + B03;
        KABB = B13 + 2.0 * B12 * B0 + 3.0 * B1 * B02 + 4.0 * B03;
        
        P[0] = P[0] + DB * C1;
        P[1] = P[1] + DB * CA;
        P[3] = P[3] + DB * CAA;
        P[6] = P[6] + DB * CAAA;
        P[2] = P[2]+ DA * CB;
        P[5] = P[5]+ DA * CBB;
        P[9] = P[9]+ DA * CBBB;
        P[4] = P[4]+ DB * (B1 * CAB + B0 * KAB);
        P[7] = P[7]+ DB * (B1 * CAAB + B0 * KAAB);
        P[8] = P[8]+ DA * (A1 * CABB + A0 * KABB);
    }
    
    P[0] = P[0] / 2.0;
    P[1] = P[1] / 6.0;
    P[3] = P[3] / 12.0;
    P[6] = P[6] / 20.0;
    P[2] = - P[2] / 6.0;
    P[5] = - P[5] / 12.0;
    P[9] = - P[9] / 20.0;
    P[4] = P[4] / 24.0;
    P[7] = P[7] / 60.0;
    P[8] = - P[8] / 60.0;
    
}

//
// get the volume, centroid and normalized inertia tensor (density = 1.0)
//

void VolumeInfoCalcu(double VolIntZeroth, double VolIntFirst[3], double VolIntSecond[6], POLYHEDRON &p)
{
    p.Volume = VolIntZeroth;

    vectorSet(p.Centroid, VolIntFirst[0]/VolIntZeroth, \
                          VolIntFirst[1]/VolIntZeroth, \
                          VolIntFirst[2]/VolIntZeroth);
    
    p.InertiaTensor[0][0] = VolIntSecond[1] + VolIntSecond[2];
    p.InertiaTensor[1][1] = VolIntSecond[2] + VolIntSecond[0];
    p.InertiaTensor[2][2] = VolIntSecond[0] + VolIntSecond[1];
    p.InertiaTensor[0][1] = - VolIntSecond[3];
    p.InertiaTensor[1][0] = - VolIntSecond[3];
    p.InertiaTensor[1][2] = - VolIntSecond[4];
    p.InertiaTensor[2][1] = - VolIntSecond[4];
    p.InertiaTensor[0][2] = - VolIntSecond[5];
    p.InertiaTensor[2][0] = - VolIntSecond[5];
    
}
                     

File: mesh/FortranPolyhedron/preload/volint.f90
!****************************************************************************
!
!  subroutine volint: 
!
!  input: 
!
!  output: 
!
!****************************************************************************

	subroutine volint()

	use global

	implicit none

	call compvolumeint

	mass=density*T0

	rc0=T1/T0

	Jmx0(1,1)=density*(T2(2)+T2(3))
	Jmx0(2,2)=density*(T2(3)+T2(1))
	Jmx0(3,3)=density*(T2(1)+T2(2))
	Jmx0(1,2)=-density*T2(4)
	Jmx0(2,1)=-density*T2(4)
	Jmx0(2,3)=-density*T2(5)
	Jmx0(3,2)=-density*T2(5)
	Jmx0(1,3)=-density*T2(6)
	Jmx0(3,1)=-density*T2(6)

	return

	end subroutine volint

!****************************************************************************
!
!  subroutine compvolumeint: 
!
!  input: 
!
!  output: 
!
!****************************************************************************

	subroutine compvolumeint()

	use global

	implicit none

	integer :: i,A,B,C

	real(8) :: nx,ny,nz,tmp

	real(8) :: wfs(N_f),nfs(N_f,3), F(12)!F=(Fa,Fb,Fc,Faa,Fbb,Fcc,Faaa,Fbbb,Fccc,Faab,Fbbc,Fcca)

	T0=0.0d0
	T1=(/0.0d0, 0.0d0, 0.0d0/)
	T2=(/0.0d0, 0.0d0, 0.0d0, 0.0d0, 0.0d0, 0.0d0/)

	call compwfnf(wfs,nfs)

	do i=1,N_f

	  nx=abs(nfs(i,1))
	  ny=abs(nfs(i,2))
	  nz=abs(nfs(i,3))
	  if ((nx>ny).and.(nx>nz)) then 
	    C=1
	  else if (ny>nz) then 
	    C=2
	  else 
	    C=3
	  end if
	  A=mod(C,3)+1
	  B=mod(A,3)+1

	  call compfaceint(wfs(i),nfs(i,:),i,A,B,C,F)

	  if (A==1) then
	    tmp=F(1)
	  else if (B==1) then
	    tmp=F(2)
	  else
	    tmp=F(3)
	  end if

	  T0=T0+nfs(i,1)*tmp

	  T1(A)=T1(A)+nfs(i,A)*F(4)
	  T1(B)=T1(B)+nfs(i,B)*F(5)
	  T1(C)=T1(C)+nfs(i,C)*F(6)
	  T2(A)=T2(A)+nfs(i,A)*F(7)
	  T2(B)=T2(B)+nfs(i,B)*F(8)
	  T2(C)=T2(C)+nfs(i,C)*F(9)
	  T2(3+A)=T2(3+A)+nfs(i,A)*F(10)
	  T2(3+B)=T2(3+B)+nfs(i,B)*F(11)
	  T2(3+C)=T2(3+C)+nfs(i,C)*F(12)

	end do

	T1=T1/2.0d0
	T2(1:3)=T2(1:3)/3.0d0
	T2(4:6)=T2(4:6)/2.0d0

	return

	end subroutine compvolumeint

!****************************************************************************
!
!  subroutine compwfnf: 
!
!  input: 
!
!  output: 
!
!****************************************************************************

	subroutine compwfnf(wfs,nfs)

	use global

	implicit none

	integer :: i

	real(8) :: wfs(N_f), nfs(N_f,3),ra(3),rb(3),rc(3),tmp(3)

	do i=1,N_f
	  
	  ra=v_data(f_data(i,1),:)
	  rb=v_data(f_data(i,2),:)
	  rc=v_data(f_data(i,3),:)
	  call cross(rb-ra,rc-rb,tmp)
	  tmp=tmp/sqrt(dot_product(tmp,tmp))
	  wfs(i)=-dot_product(ra,tmp)
	  nfs(i,:)=tmp
	  
	end do

	return

	end subroutine compwfnf

!****************************************************************************
!
!  subroutine compfaceint: 
!
!  input: 
!
!  output: 
!
!****************************************************************************

	subroutine compfaceint(w,nf,id,A,B,C,F)
    use, intrinsic :: iso_fortran_env
	implicit none

	integer :: A,B,C,id

	real(8) :: w,nf(3),F(12),P(10) !p=(P1,Pa,Pb,Paa,Pab,Pbb,Paaa,Paab,Pabb,Pbbb)

	real(8) :: k(4)

	call compprojectionint(A,B,id,P)

	k(1)=1.0d0/nf(C)
	k(2)=k(1)*k(1)
	k(3)=k(2)*k(1)
	k(4)=k(3)*k(1)

	F(1)=k(1)*P(2)
	F(2)=k(1)*P(3)
	F(3)=-k(2)*(nf(A)*P(2)+nf(B)*P(3)+w*P(1))

	F(4)=k(1)*P(4)
	F(5)=k(1)*P(6)
	F(6)=k(3)*((nf(A)**2)*P(4)+2.0d0*nf(A)*nf(B)*P(5)&
		 +(nf(B)**2)*P(6)+w*(2.0d0*(nf(A)*P(2)+nf(B)*P(3))+w*P(1)))

	F(7)=k(1)*P(7)
	F(8)=k(1)*P(10)
	F(9)=-k(4)*((nf(A)**3)*P(7)+3.0d0*(nf(A)**2)*nf(B)*P(8)& 
	     +3.0d0*nf(A)*(nf(B)**2)*P(9)+(nf(B)**3)*P(10)&
	     +3.0d0*w*((nf(A)**2)*P(4)+2.0d0*nf(A)*nf(B)*P(5)+(nf(B)**2)*P(6))&
	     +w*w*(3.0d0*(nf(A)*P(2)+nf(B)*P(3))+w*P(1)))

	F(10)=k(1)*P(8)
	F(11)=-k(2)*(nf(A)*P(9)+nf(B)*P(10)+w*P(6))
	F(12)=k(3)*((nf(A)**2)*P(7)+2.0d0*nf(A)*nf(B)*P(8)+(nf(B)**2)*P(9)&
	      +w*(2.0d0*(nf(A)*P(4)+nf(B)*P(5))+w*P(2)))

	return

	end subroutine compfaceint

!****************************************************************************
!
!  subroutine compprojectionint: 
!
!  input: 
!
!  output: 
!
!****************************************************************************

	subroutine compprojectionint(A,B,id,P)

	use global

	implicit none

	integer :: A,B,id,i
	real(8) :: P(10)
	real(8) :: a0,a1,da
	real(8) :: b0,b1,db
	real(8) :: a0_2,a0_3,a0_4,b0_2,b0_3,b0_4
	real(8) :: a1_2,a1_3,b1_2,b1_3
	real(8) :: C1,Ca,Caa,Caaa,Cb,Cbb,Cbbb
	real(8) :: Cab,Kab,Caab,Kaab,Cabb,Kabb

	P=(/0.0d0, 0.0d0, 0.0d0, 0.0d0, 0.0d0, 0.0d0, 0.0d0, 0.0d0, 0.0d0, 0.0d0/)

	do i=1,3

	  a0=v_data(f_data(id,i),A)
      b0=v_data(f_data(id,i),B)
      a1=v_data(f_data(id,mod(i,3)+1),A)
      b1=v_data(f_data(id,mod(i,3)+1),B)
      da=a1-a0
      db=b1-b0

      a0_2=a0*a0
	  a0_3=a0_2*a0
	  a0_4=a0_3*a0

      b0_2=b0*b0
	  b0_3=b0_2*b0
	  b0_4=b0_3*b0

      a1_2=a1*a1
	  a1_3=a1_2*a1 
      b1_2=b1*b1
	  b1_3=b1_2*b1

      C1=a1+a0
      Ca=a1*C1+a0_2
	  Caa=a1*Ca+a0_3
	  Caaa=a1*Caa+a0_4
      Cb=b1*(b1+b0)+b0_2
	  Cbb=b1*Cb+b0_3
	  Cbbb=b1*Cbb+b0_4
      Cab=3.0d0*a1_2+2.0d0*a1*a0+a0_2
	  Kab=a1_2+2.0d0*a1*a0+3.0d0*a0_2
      Caab=a0*Cab+4.0d0*a1_3
	  Kaab=a1*Kab+4.0d0*a0_3
      Cabb=4.0d0*b1_3+3.0d0*b1_2*b0+2.0d0*b1*b0_2+b0_3
      Kabb=b1_3+2.0d0*b1_2*b0+3.0d0*b1*b0_2+4.0d0*b0_3

!p=(P1,Pa,Pb,Paa,Pab,Pbb,Paaa,Paab,Pabb,Pbbb)

      P(1)=P(1)+db*C1
      P(2)=P(2)+db*Ca
      P(4)=P(4)+db*Caa
      P(7)=P(7)+db*Caaa
      P(3)=P(3)+da*Cb
      P(6)=P(6)+da*Cbb
      P(10)=P(10)+da*Cbbb
      P(5)=P(5)+db*(b1*Cab+b0*Kab)
      P(8)=P(8)+db*(b1*Caab+b0*Kaab)
      P(9)=P(9)+da*(a1*Cabb+a0*Kabb)

	end do

	P(1)=P(1)/2.0d0
    P(2)=P(2)/6.0d0
    P(4)=P(4)/12.0d0
    P(7)=P(7)/20.0d0
    P(3)=-P(3)/6.0d0
    P(6)=-P(6)/12.0d0
    P(10)=-P(10)/20.0d0
    P(5)=P(5)/24.0d0
    P(8)=P(8)/60.0d0
    P(9)=-P(9)/60.0d0

	return

	end subroutine compprojectionint